src/mon/AuthMonitor.cc

```diff
@@ -1795,6 +1795,47 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
       + " tag " + pg_pool_t::APPLICATION_NAME_CEPHFS
       + " data=" + filesystem;
 
+    vector<string> newcaps = {"mon", mon_cap_string, "osd", osd_cap_string,
+			      "mds", mds_cap_string};
+
+    EntityAuth e_auth;
+    if (mon.key_server.get_auth(entity, e_auth)) {
+      for (auto it = newcaps.begin(); it != newcaps.end(); it += 2) {
+	string& cap_entity = *it;
+        if (entity_auth.caps.count(cap_entity) == 0) {
+          ss << entity << " already has fs capabilities that differ from "
+             << "those supplied. To generate a new auth key for " << entity
+             << ", first remove " << entity << " from configuration files, "
+             << "execute 'ceph auth rm " << entity << "', then execute this "
+             << "command again.";
+          err = -EINVAL;
+          goto done;
+        }
+      }
+
+      int rv = _gen_wanted_caps(e_auth, newcaps, &ss);
+      if (rv == 2) {
```

Comment:
This should check the enumerated value.

Reply:
By defining it in the header or even in the cc source in a scope that accessible to other class methods.

Reply:
You didn't mention this before. I was under impression that we want to use enum only inside _gen_wanted_caps().

Reply:
Sorry, it was pretty obvious to me since if you define an enum for a function to return that enum, the callers should check against those enumerated value types.

Reply:
> I've made this change. PTAL.

 IMHO this getting much more complicated than using simple int. Especially, because this enum is not used beyond the very limited context of _gen_wanted_caps().